### PR TITLE
Support export P11PBE key in FIPS mode

### PR DESCRIPTION
**Pull request by WilburZjh**
PBE related crypto services are now supported in the SunPKCS11 provider. Therefore, in FIPS mode, users can directly utilize those services to generate a PBE secret key.

Currently, we only support to export RSA/EC private key and AES secret key. We also need to provide the support to export P11PBE key in FIPS mode.

### CODE REVIEW

**Comment by WilburZjh** [comment]
@jasonkatonica FYI

**Comment by WilburZjh** [comment]
Hi @keithc-ca , the tests and build are finished, please help to review

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java [hunk 1]
```diff
@@ -381,6 +383,18 @@ static SecretKey pbeKey(Session session, long keyID, String algorithm,
             new CK_ATTRIBUTE(CKA_SENSITIVE),
             new CK_ATTRIBUTE(CKA_EXTRACTABLE),
         });
+        if ((SunPKCS11.mysunpkcs11 != null) && attrs[0].getBoolean() || attrs[1].getBoolean() || (attrs[2].getBoolean() == false)) {
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java:386]: This line should be folded, and parentheses are missing here: I think around the `||` portion:
```java
        if ((SunPKCS11.mysunpkcs11 != null)
            && (attrs[0].getBoolean()
                || attrs[1].getBoolean()
                || (attrs[2].getBoolean() == false))
        ) {
```
See also lines 358-361.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java [hunk 2]
```diff
@@ -108,6 +108,8 @@ abstract class P11Key implements Key, Length {
 
     private static final boolean DISABLE_NATIVE_KEYS_EXTRACTION;
 
+    private static boolean isPlainPBEKeySupportInFIPS = false;
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java:111]: A static field is not appropriate for this purpose; `key != null` is a suitable replacement.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java [hunk 3]
```diff
@@ -521,14 +535,23 @@ private static class P11SecretKey extends P11Key implements SecretKey {
...
+        P11SecretKey(Session session, long keyID, String algorithm,
+                int keyLength, CK_ATTRIBUTE[] attrs, SecretKey key) {
+            super(SECRET, session, keyID, algorithm, keyLength, attrs);
+            this.key = key;
         }
 
         public String getFormat() {
             token.ensureValid();
-            if (sensitive || !extractable || (isNSS && tokenObject)) {
+            if (!isPlainPBEKeySupportInFIPS && (sensitive || !extractable || (isNSS && tokenObject))) {
```
Comment (keithc-ca) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java:554]:
```java
            if ((key == null) && (sensitive || !extractable || (isNSS && tokenObject))) {
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by keithc-ca** [comment]
We don't want lines of commit messages to be too long (i.e. no more than 72 characters), but having them too short isn't ideal either. Please reformat the message to use longer (but not too long) lines.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java (hunk 1 above)
Comment (WilburZjh) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java]: Updated

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java (hunk 2 above)
Comment (WilburZjh) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java]: Updated

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java [hunk 4]
```diff
@@ -381,6 +383,18 @@ static SecretKey pbeKey(Session session, long keyID, String algorithm,
             new CK_ATTRIBUTE(CKA_SENSITIVE),
             new CK_ATTRIBUTE(CKA_EXTRACTABLE),
         });
+        if ((SunPKCS11.mysunpkcs11 != null) && attrs[0].getBoolean() || attrs[1].getBoolean() || (attrs[2].getBoolean() == false)) {
+            try {
+                byte[] key = SunPKCS11.mysunpkcs11.exportKey(session.id(), attrs, keyID);
+                SecretKey secretKey = new SecretKeySpec(key, algorithm);
+                return new P11PBEKey(session, keyID, algorithm, keyLength, attrs, password, salt, iterationCount, secretKey);
+            } catch (PKCS11Exception e) {
+                // Attempt failed, create a pbeKey object.
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java]: Updated

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java (hunk 3 above)
Comment (WilburZjh) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java]: Updated

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java [hunk 5]
```diff
@@ -541,6 +564,10 @@ byte[] getEncodedInternal() {
                 return null;
             }
 
+            if (isPlainPBEKeySupportInFIPS) {
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java]: Updated

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java [hunk 6]
```diff
@@ -577,13 +604,28 @@ static final class P11PBEKey extends P11SecretKey
         private char[] password;
         private final byte[] salt;
         private final int iterationCount;
+        private final SecretKey key;
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java]: Updated

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java [hunk 7]
```diff
@@ -381,6 +381,22 @@ static SecretKey pbeKey(Session session, long keyID, String algorithm,
             new CK_ATTRIBUTE(CKA_SENSITIVE),
             new CK_ATTRIBUTE(CKA_EXTRACTABLE),
         });
+        if ((SunPKCS11.mysunpkcs11 != null) 
+            && (attrs[0].getBoolean() 
+                || attrs[1].getBoolean()
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java]: Updated

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java [hunk 8]
```diff
@@ -381,6 +381,22 @@ static SecretKey pbeKey(Session session, long keyID, String algorithm,
...
+                || (attrs[2].getBoolean() == false))
+        ) {
+            try {
+                byte[] key = SunPKCS11.mysunpkcs11.exportKey(session.id(), attrs, keyID);
+                SecretKey secretKey = new SecretKeySpec(key, algorithm);
+                return new P11PBEKey(session, keyID, algorithm, keyLength, attrs, password, salt, iterationCount, secretKey);
+            } catch (PKCS11Exception e) {
+                // Attempt failed, create a regular P11PBEKey.
+                if (debug != null) {
+                    debug.println("Attempt failed, creating a pbeKey object for " + algorithm);
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java]: Updated

**Comment by WilburZjh** [comment]
> We don't want lines of commit messages to be too long (i.e. no more than 72 characters), but having them too short isn't ideal either. Please reformat the message to use longer (but not too long) lines.

Updated

**Comment by keithc-ca** [comment]
Jenkins test sanity amac jdknext

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**